contents/docs/runbook/services/plugin-server/ingestion-lag.mdx

```diff
@@ -0,0 +1,211 @@
+
+
+### Dealing with hot partitions
+
+If you see from the Kafka consumer lag graph ([US Cloud link](https://metrics.posthog.net/d/27BWK5nGk/kafka-msk-overview?orgId=1&viewPanel=23)) that our lag seems to stem from only one or a few partitions, that's likely because we have a hot partition. 
+
+We partition the `events_plugin_ingestion` topic by `<team_id>:<distinct_id>`, meaning we're likely getting a burst of events for a given person in a short period of time, causing a hot partition. 
+
+There are a few possible mitigation steps for this:
+
+1. Wait. If the burst is temporary we should recover in time.
+2. Discover the problematic IDs and force events being sent to them to be randomly partitioned.
+
+In order to do #2, you should first try to discover what IDs could be causing problems. Here's a useful query to run on ClickHouse:
+
+```sql
+SELECT
+    team_id,
+    distinct_id,
+    count(*) AS c
+FROM events
+WHERE 
+    timestamp > now() - INTERVAL 1 HOUR AND
+    timestamp < now()
+GROUP BY
+    team_id,
+    distinct_id
+ORDER BY c DESC
+LIMIT 20
+```
+
+If an ID stands out, you can do some further digging with:
+
+```sql
+SELECT 
+    toStartOfMinute(timestamp) ts,
+    count(*)
+FROM events 
+WHERE 
+    team_id=<team_id> AND 
+    distinct_id=<distinct_id> AND 
+    timestamp > now() - INTERVAL 1 DAY AND
+    timestamp < now() 
+GROUP BY ts
+```
+
+Running the above query on Metabase gives you a very useful visualization, where you can see if events for this ID have increased significantly recently or if the volume is always high.
+
+Often times this graph will give you a clear indication of if an ID caused the hot partition or not.
+
+If you've established that there's likely an ID causing a hot partition, it's worth peeking into the properties for the person it's associated to:
+
+```sql
+WITH p AS ( 
+    SELECT 
+        id, 
+        properties 
+    FROM person
+    WHERE 
+        team_id=<team_id>
+),
+pdi AS (
+    SELECT person_id 
+    FROM person_distinct_id2
+    WHERE
+        team_id=<team_id> AND 
+        distinct_id=<distinct_id>
+)
+SELECT p.properties 
+FROM p JOIN pdi
+ON p.id = pdi.person_id
+```
+
+If it looks like the properties are all / mostly generic (e.g. initial referrers, OS, properties PostHog sets by default) **or** the hot partition problem is **very serious** (a lot of lag), you can then [copy the approach from this PR](https://github.com/PostHog/posthog-cloud-infra/pull/806) to force the events for the ID to be randomly partitioned.
+
+We check the person properties because if a lot of custom properties are being used random partitioning may cause properties to get inconsistent values.
+
+### Dropping and recreating ClickHouse "ingestion tables"
+
+If you've established that ClickHouse is not ingesting data correctly (via Kafka consumer lag graphs, ClickHouse logs, etc), then it could be worth recreating the relevant Kafka table + materialized view pairs that are responsible for ingesting data.
```

Comment:
what's the impact of recreating these tables? Does this one deserve a warning too?

Reply:
It's fine to do this